### PR TITLE
CDAP-15455. Support Upsert and Update operations for SFDC Sink

### PR DESCRIPTION
**Pull request by aonischuk**
Without the change only insert is supported.

_How new operations work in Salesforce:_
**Upsert** - Salesforce requires user to provide an external id field name. This field is used as basis for upsert (Salesforce will decide if objects are the same using it).

Id field can be used for that (which is present for all Salesforce sObjects). Also user can create a custom field and checkmark it as "External Id" via Salesforce UI.

**Update** - Specifying an external id field is not supported. For updates Salesforce will always use 'Id' field as basis.

### CODE REVIEW

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 1]
```diff
@@ -182,20 +268,28 @@ private void validateSchema(Schema schema) {
...
+    PartnerConnection partnerConnection = null;
+    SObjectDescriptor sObjectDescriptor = null;
+    try {
+      partnerConnection = new PartnerConnection(Authenticator.createConnectorConfig(credentials));
+      sObjectDescriptor = SObjectDescriptor.fromName(this.getSObject(),
+                                                     this.getAuthenticatorCredentials());
+    } catch (ConnectionException e) {
+      throw new InvalidStageException("There was issue communicating with Salesforce", e);
+    }
+    return new SObjectsDescribeResult(partnerConnection, sObjectDescriptor.getAllParentObjects());
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: you don't have to return outside of try block. That means you don't have to init all variables with null. Just move all declarations and return into try block.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java [hunk 2]
```diff
@@ -182,20 +268,28 @@ private void validateSchema(Schema schema) {
...
     }
     return creatableSObjectFields;
   }
+
+  private SObjectsDescribeResult getSObjectDescribeResult() {
+    AuthenticatorCredentials credentials = this.getAuthenticatorCredentials();
+    PartnerConnection partnerConnection = null;
+    SObjectDescriptor sObjectDescriptor = null;
+    try {
+      partnerConnection = new PartnerConnection(Authenticator.createConnectorConfig(credentials));
```
Comment (mlozbin-cybervisiontech) [inline src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: it seems strange to do IO as a side effect in configuration class

**Review by mlozbin-cybervisiontech — CHANGES_REQUESTED** [review]

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 2 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: This is validation logic. It works based on information received from Salesforce server. We validate are the credentials correct, are fields present, are they editable etc. etc.
This is how it works in all other SF plugins as well.

File: src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java (hunk 1 above)
Comment (aonischuk) [reply src/main/java/io/cdap/plugin/salesforce/plugin/sink/batch/SalesforceSinkConfig.java]: done

**Review by mlozbin-cybervisiontech — APPROVED** [review]

**Pull request closed without merging**